app/precompiled.py

```diff
@@ -517,7 +517,10 @@ def _colour_no_print_areas_of_single_page_pdf_in_red(src_pdf, is_first_page):
     :param BytesIO src_pdf: A file-like representing a single page pdf
     :param bool is_first_page: true if we should overlay the address block red area too.
     """
-    pdf = PdfFileReader(src_pdf)
+    try:
+        pdf = PdfFileReader(src_pdf)
+    except PdfReadError as e:
```

Comment:
I can see that the `binascii.Error` will no longer be caught by a specific handler and instead be handled by the default exception handler.

However, I can see that we will be catching `PdfReadError`s here and then raising an `InvalidRequest`.

Why do we have different behaviour for the two different exceptions? Why not catch either of them here? Or leave both of them to go uncaught?

Reply:
Just to check I understood, you wanted to leave both of them uncaught but you found an existing test for the `PdfReadError` and decided that provided enough reason for you to explicitly catch it, rather than say deleting the test?

And then also you say you wanted to leave both of them uncaught but in the sentence above you said that in general we should avoid global exception handlers unless the exception is global and these aren't. That seems to say to me to be contradictory (you say you want them uncaught but before it said that these are not good candidates for the global exception handler) or maybe I've misunderstood your comment...

Reply:
> you say you want them uncaught but before it said that these are not good candidates for the global exception handler

Ah, I should have been clearer when I said "uncaught".

I mean that they should be caught by the general global exception handler, not the specific global exception handlers I'm removing in the first commit. It's really another instance of [your comment about the 400 handler](https://github.com/alphagov/notifications-template-preview/pull/637#discussion_r815916397): these handlers have been downgrading all  `PdfReadError` and `binascii.Error` exceptions to a 400 response / warning log, but sometimes the exception may actually represent a bug and we'd miss it.

> decided that provided enough reason for you to explicitly catch it, rather than say deleting the test?

The test is evidence that we need to handle the exception at a specific place in the code, in addition to any other places where we handle it already. But there's no evidence that these exceptions are always benign. The fact that we don't seem to get any of them in Production means there's little risk of removing the global handlers that currently downgrade the logs for all such exceptions - irrespective of how they occurred.

Is that making more sense now?

Reply:
I think it is making a bit more sense. I don't know if I understand it well enough to reexplain it to someone else but I think that is OK.

Thank you for taking the time on this.